README.md
<span><img src=".\Assets\img\powershell_500px.png" height="10%" width="10%" align=left /><h1>PSClearHost</h1><img src=".\Assets\logo.png" height="15%" width="15%" align=right /></span>

***

> A PowerShell Module that adds awesome `Clear-Host` animations for your shell/terminal!

## Links

- [Source Code](https://github.com/jimbrig.com/PSClearHost)
- [Latest Release - v1.0.0](https://github.com/jimbrig/PSClearHost/releases/tag/v1.0.0)
- [Published PowerShell Gallery Module](https://www.powershellgallery.com/packages/PSClearHost/1.0.0)
- [Documentation Site](https://docs.jimbrig.com/PSClearHost/)

*View the repo's [Changelog](CHANGELOG.md) for details on the progression of the codebase over time.*

## Badges

[![PowerShell Gallery Version](https://img.shields.io/powershellgallery/v/PSClearHost?color=0092ff&label=PowerShell%20Gallery&logoColor=0092ff)](https://www.powershellgallery.com/packages/PSClearHost/1.0.0)

[![Publish GitHub NuGet Package](https://github.com/jimbrig/PSClearHost/actions/workflows/ghpkg.yml/badge.svg)](https://github.com/jimbrig/PSClearHost/actions/workflows/ghpkg.yml)

[![Test Module](https://github.com/jimbrig/PSClearHost/actions/workflows/test.yml/badge.svg)](https://github.com/jimbrig/PSClearHost/actions/workflows/test.yml)
[![Build Module](https://github.com/jimbrig/PSClearHost/actions/workflows/build.yml/badge.svg?branch=develop)](https://github.com/jimbrig/PSClearHost/actions/workflows/build.yml)
[![Publish Module](https://github.com/jimbrig/PSClearHost/actions/workflows/publish.yml/badge.svg)](https://github.com/jimbrig/PSClearHost/actions/workflows/publish.yml)
[![Publish Docs Site](https://github.com/jimbrig/PSClearHost/actions/workflows/mkdocs.yml/badge.svg)](https://github.com/jimbrig/PSClearHost/actions/workflows/mkdocs.yml)
[![Changelog](https://github.com/jimbrig/PSClearHost/actions/workflows/changelog.yml/badge.svg)](https://github.com/jimbrig/PSClearHost/actions/workflows/changelog.yml)

## Installation

```powershell
Install-Module PSClearHost -Scope CurrentUser
```

## Usage

```powershell
# Clear host using the "Bricks" mode
Clear-HostRaindrop -Mode Bricks -Speed 5.0

# Clear host using the "Falling" mode
Clear-HostRaindrop -Mode Falling -Speed 2.0

# Clear host using the "Flipping" mode
Clear-HostRaindrop -Mode Flipping -Speed 2.0
```

## Demos

Click on a demonstration video below to see the module in action!

### Bricks Mode

https://user-images.githubusercontent.com/32652297/200099398-29dd1d6c-3ffa-4ec6-9833-7bbfbcdf2fc7.mp4

### Falling Mode

https://user-images.githubusercontent.com/32652297/200099413-5400a1f9-c88d-4e60-b8ef-c07013f73c98.mp4

### Flipping Mode

https://user-images.githubusercontent.com/32652297/200099418-22c6cc4c-ca8d-4dcf-8328-d451b450456a.mp4


CHANGELOG.md
# Changelog
*All notable changes to this project will be documented in this file.*


### Bug Fixes

- Fix mkdocs config

### Documentation

- Tweak README.md
- Fix demos in readme
- Add docs for publishing
- Move Publish.md
- Add ghpackage nuget badge to readme

### Features

- Add ghpkg.yml workflow
- Add publish nuget ghpackage workflow

## [1.0.0] - 2022-11-05

### Bug Fixes

- Bug fixes

### Documentation

- Prep changelog for release notes
- Add mp4 videos for README demos
- Add various assets
- Update all documentation resources
- Update function docs

### Features

- Add devcontainer
- Add editorconfig
- Add .gitattributes

### Testing

- TestResults
- Run tests

***
*Changelog generated by [git-cliff](https://github.com/orhun/git-cliff).*


Tests/RunTests.ps1
#Requires -Modules @{ ModuleName="Pester"; ModuleVersion="5.0.0"}

$config = New-PesterConfiguration
$config.Run.PassThru = $true
$config.Run.Path = $PSScriptRoot

Invoke-Pester -Configuration $config


Tests/Help.Tests.ps1
# Taken with love from @juneb_get_help (https://raw.githubusercontent.com/juneb/PesterTDD/master/Module.Help.Tests.ps1)

BeforeDiscovery {

    function global:FilterOutCommonParams {
        param ($Params)
        $commonParams = @(
            'Debug', 'ErrorAction', 'ErrorVariable', 'InformationAction', 'InformationVariable',
            'OutBuffer', 'OutVariable', 'PipelineVariable', 'Verbose', 'WarningAction',
            'WarningVariable', 'Confirm', 'Whatif'
        )
        $params | Where-Object { $_.Name -notin $commonParams } | Sort-Object -Property Name -Unique
    }

    $manifest             = Import-PowerShellDataFile -Path $env:BHPSModuleManifest
    $outputDir            = Join-Path -Path $env:BHProjectPath -ChildPath 'Output'
    $outputModDir         = Join-Path -Path $outputDir -ChildPath $env:BHProjectName
    $outputModVerDir      = Join-Path -Path $outputModDir -ChildPath $manifest.ModuleVersion
    $outputModVerManifest = Join-Path -Path $outputModVerDir -ChildPath "$($env:BHProjectName).psd1"

    # Get module commands
    # Remove all versions of the module from the session. Pester can't handle multiple versions.
    Get-Module $env:BHProjectName | Remove-Module -Force -ErrorAction Ignore
    Import-Module -Name $outputModVerManifest -Verbose:$false -ErrorAction Stop
    $params = @{
        Module      = (Get-Module $env:BHProjectName)
        CommandType = [System.Management.Automation.CommandTypes[]]'Cmdlet, Function' # Not alias
    }
    if ($PSVersionTable.PSVersion.Major -lt 6) {
        $params.CommandType[0] += 'Workflow'
    }
    $commands = Get-Command @params

    ## When testing help, remember that help is cached at the beginning of each session.
    ## To test, restart session.
}

Describe "Test help for <_.Name>" -ForEach $commands {

    BeforeDiscovery {
        # Get command help, parameters, and links
        $command               = $_
        $commandHelp           = Get-Help $command.Name -ErrorAction SilentlyContinue
        $commandParameters     = global:FilterOutCommonParams -Params $command.ParameterSets.Parameters
        $commandParameterNames = $commandParameters.Name
        $helpLinks             = $commandHelp.relatedLinks.navigationLink.uri
    }

    BeforeAll {
        # These vars are needed in both discovery and test phases so we need to duplicate them here
        $command                = $_
        $commandName            = $_.Name
        $commandHelp            = Get-Help $command.Name -ErrorAction SilentlyContinue
        $commandParameters      = global:FilterOutCommonParams -Params $command.ParameterSets.Parameters
        $commandParameterNames  = $commandParameters.Name
        $helpParameters         = global:FilterOutCommonParams -Params $commandHelp.Parameters.Parameter
        $helpParameterNames     = $helpParameters.Name
    }

    # If help is not found, synopsis in auto-generated help is the syntax diagram
    It 'Help is not auto-generated' {
        $commandHelp.Synopsis | Should -Not -BeLike '*`[`<CommonParameters`>`]*'
    }

    # Should be a description for every function
    It "Has description" {
        $commandHelp.Description | Should -Not -BeNullOrEmpty
    }

    # Should be at least one example
    It "Has example code" {
        ($commandHelp.Examples.Example | Select-Object -First 1).Code | Should -Not -BeNullOrEmpty
    }

    # Should be at least one example description
    It "Has example help" {
        ($commandHelp.Examples.Example.Remarks | Select-Object -First 1).Text | Should -Not -BeNullOrEmpty
    }

    It "Help link <_> is valid" -ForEach $helpLinks {
        (Invoke-WebRequest -Uri $_ -UseBasicParsing).StatusCode | Should -Be '200'
    }

    Context "Parameter <_.Name>" -Foreach $commandParameters {

        BeforeAll {
            $parameter         = $_
            $parameterName     = $parameter.Name
            $parameterHelp     = $commandHelp.parameters.parameter | Where-Object Name -eq $parameterName
            $parameterHelpType = if ($parameterHelp.ParameterValue) { $parameterHelp.ParameterValue.Trim() }
        }

        # Should be a description for every parameter
        It "Has description" {
            $parameterHelp.Description.Text | Should -Not -BeNullOrEmpty
        }

        # Required value in Help should match IsMandatory property of parameter
        It "Has correct [mandatory] value" {
            $codeMandatory = $_.IsMandatory.toString()
            $parameterHelp.Required | Should -Be $codeMandatory
        }

        # Parameter type in help should match code
        It "Has correct parameter type" {
            $parameterHelpType | Should -Be $parameter.ParameterType.Name
        }
    }

    Context "Test <_> help parameter help for <commandName>" -Foreach $helpParameterNames {

        # Shouldn't find extra parameters in help.
        It "finds help parameter in code: <_>" {
            $_ -in $parameterNames | Should -Be $true
        }
    }
}

Tests/Meta.Tests.ps1
BeforeAll {

    Set-StrictMode -Version latest

    # Make sure MetaFixers.psm1 is loaded - it contains Get-TextFilesList
    Import-Module -Name (Join-Path -Path $PSScriptRoot -ChildPath 'MetaFixers.psm1') -Verbose:$false -Force

    $projectRoot = $ENV:BHProjectPath
    if (-not $projectRoot) {
        $projectRoot = $PSScriptRoot
    }

    $allTextFiles      = Get-TextFilesList $projectRoot
    $unicodeFilesCount = 0
    $totalTabsCount    = 0
    foreach ($textFile in $allTextFiles) {
        if (Test-FileUnicode $textFile) {
            $unicodeFilesCount++
            Write-Warning (
                "File $($textFile.FullName) contains 0x00 bytes." +
                " It probably uses Unicode/UTF-16 and needs to be converted to UTF-8." +
                " Use Fixer 'Get-UnicodeFilesList `$pwd | ConvertTo-UTF8'."
            )
        }
        $unicodeFilesCount | Should -Be 0

        $fileName = $textFile.FullName
        (Get-Content $fileName -Raw) | Select-String "`t" | Foreach-Object {
            Write-Warning (
                "There are tabs in $fileName." +
                " Use Fixer 'Get-TextFilesList `$pwd | ConvertTo-SpaceIndentation'."
            )
            $totalTabsCount++
        }
    }
}

Describe 'Text files formatting' {
    Context 'File encoding' {
        It "No text file uses Unicode/UTF-16 encoding" {
            $unicodeFilesCount | Should -Be 0
        }
    }

    Context 'Indentations' {
        It "No text file use tabs for indentations" {
            $totalTabsCount | Should -Be 0
        }
    }
}

build.ps1
[cmdletbinding(DefaultParameterSetName = 'Task')]
param(
    # Build task(s) to execute
    [parameter(ParameterSetName = 'task', position = 0)]
    [string[]]$Task = 'default',

    # Bootstrap dependencies
    [switch]$Bootstrap,

    # List available build tasks
    [parameter(ParameterSetName = 'Help')]
    [switch]$Help,

    [pscredential]$PSGalleryApiKey
)

$ErrorActionPreference = 'Stop'

# Bootstrap dependencies
if ($Bootstrap.IsPresent) {
    Get-PackageProvider -Name Nuget -ForceBootstrap | Out-Null
    Set-PSRepository -Name PSGallery -InstallationPolicy Trusted
    if (-not (Get-Module -Name PSDepend -ListAvailable)) {
        Install-Module -Name PSDepend -Repository PSGallery -Scope CurrentUser
    }
    Invoke-PSDepend -Path './requirements.psd1' -Install -Import -Force -WarningAction SilentlyContinue
}

# Execute psake task(s)
$psakeFile = './psakeFile.ps1'
if ($PSCmdlet.ParameterSetName -eq 'Help') {
    Get-PSakeScriptTasks -buildFile $psakeFile  |
        Format-Table -Property Name, Description, Alias, DependsOn
} else {
    Set-BuildEnvironment -Force
    $parameters = @{}
    if ($PSGalleryApiKey) {
        $parameters['galleryApiKey'] = $PSGalleryApiKey
    }
    Invoke-psake -buildFile $psakeFile -taskList $Task -nologo -parameters $parameters
    exit ( [int]( -not $psake.build_success ) )
}

psakeFile.ps1
properties {
    # Set this to $true to create a module with a monolithic PSM1
    $PSBPreference.Build.CompileModule = $false
    $PSBPreference.Help.DefaultLocale = 'en-US'
    $PSBPreference.Test.OutputFile = 'out/testResults.xml'
    $PSBPreference.Build.CopyDirectories = @()

    # Publish settings
    if ($galleryApiKey) {
        $PSBPreference.Publish.PSRepositoryApiKey = $galleryApiKey.GetNetworkCredential().password
    }
}

task Default -depends Test

task Test -FromModule PowerShellBuild -minimumVersion '0.6.1'

task InstallAct {
    if (-not (Get-Command -Name act -CommandType Application -ErrorAction SilentlyContinue)) {
        if ($IsWindows) {
            choco install act-cli
        } elseif ($IsLinux) {
            curl https://raw.githubusercontent.com/nektos/act/master/install.sh | sudo bash
        } elseIf ($IsMacOS) {
            brew install nektos/tap/act
        }
    } else {
        'act already installed'
    }
}

task TestGHAction -depends Build, InstallAct  {
    act -j test -P ubuntu-latest=nektos/act-environments-ubuntu:18.04
}


PSClearHost/Private/Bricks.ps1
using module .\Common.psm1

class Bricks
{
    static $kBallSpeed = 60
    static $kBarSpeed = 70
    static $kBarVelocityInfluenceAccelTime = 0.6
    static $kExplosionDuration = 0.6
    static $kExplosionPropagationDelayTime = [RandomDouble]::new(0.2, 0.3)

    static $foregroundColor = $null
    static $backgroundColor = $null
    static $windowWidth = 0
    static $windowHeight = 0

    static $characters = $null
    static $cellToCharacter = $null
    static $ball = $null
    static $bar = $null
    static $explosionsToAdd = $null
    static $explosions = $null

    static [void] Init($windowBuffer)
    {
        [Bricks]::foregroundColor = (Get-Host).UI.RawUI.ForegroundColor
        [Bricks]::backgroundColor = (Get-Host).UI.RawUI.BackgroundColor

        [Bricks]::CreateCharacters($windowBuffer)
        [Bricks]::CreateBall()
        [Bricks]::CreateBar()
        [Bricks]::explosionsToAdd = [System.Collections.Generic.List[PSObject]]::new()
        [Bricks]::explosions = [System.Collections.Generic.List[PSObject]]::new()
    }

    static [void] Term()
    {
        [Bricks]::characters = $null
        [Bricks]::ball = $null
        [Bricks]::bar = $null
        [Bricks]::explosions = $null
        [Bricks]::explosionsToAdd = $null
    }

    static [Boolean] Render([System.Management.Automation.Host.BufferCell[,]]$windowBuffer, [Double]$dt, [Double]$speed)
    {
        if (GetKeyHold ([ConsoleKey]::LeftArrow))
        {
            [Bricks]::bar.MoveLeft()
        }
        elseif (GetKeyHold ([ConsoleKey]::RightArrow))
        {
            [Bricks]::bar.MoveRight()
        }

        [Bricks]::bar.Update($dt)
        [Bricks]::ball.Update($dt * $speed)

        foreach ($explosion in [Bricks]::explosions)
        {
            $explosion.Update($dt * $speed)
        }
        foreach ($explosion in [Bricks]::explosionsToAdd)
        {
            [Bricks]::explosions.Add($explosion)
        }
        [Bricks]::explosionsToAdd.Clear()

        $isAllFinished = $true
        foreach ($character in [Bricks]::characters)
        {
            if (-not $character.isFinished)
            {
                $isAllFinished = $false
                break
            }
        }
        foreach ($explosion in [Bricks]::explosions)
        {
            if (-not $explosion.isFinished)
            {
                $isAllFinished = $false
                break
            }
        }

        if ($isAllFinished)
        {
            return $true
        }

        foreach ($character in [Bricks]::characters)
        {
            $character.Render($windowBuffer)
        }
        [Bricks]::ball.Render($windowBuffer)
        [Bricks]::bar.Render($windowBuffer)
        foreach ($explosion in [Bricks]::explosions)
        {
            $explosion.Render($windowBuffer)
        }

        return $false
    }

    static [void] CreateCharacters($windowBuffer)
    {
        [Bricks]::windowWidth = $windowBuffer.GetUpperBound(1)
        [Bricks]::windowHeight = $windowBuffer.GetUpperBound(0)

        [Bricks]::characters = CreateCharacters ([BricksCharacter]) $windowBuffer

        [Bricks]::cellToCharacter = New-Object 'Object[,]' ([Bricks]::windowWidth, [Bricks]::windowHeight)
        foreach ($character in [Bricks]::characters)
        {
            [Bricks]::cellToCharacter[$character.x, $character.y] = $character
            if ($character.IsTwoCell())
            {
                [Bricks]::cellToCharacter[($character.x+1), $character.y] = $character
            }
        }
    }

    static [void] CreateBall()
    {
        $x = [Int]([Bricks]::windowWidth / 2)
        $y = [Bricks]::windowHeight - 3
        [Bricks]::ball = [BricksBall]::new($x, $y)
    }

    static [void] CreateBar()
    {
        $width = 20
        $x = [Int](([Bricks]::windowWidth - $width) / 2)
        $y = [Int]([Bricks]::windowHeight - 2)
        [Bricks]::bar = [BricksBar]::new($x, $y, $width)
    }

    static [void] CreateExplosion($character, $isBackgroundColorAnimation)
    {
        [Bricks]::explosionsToAdd.Add([Explosion]::new($character, $isBackgroundColorAnimation))
    }

    static [Object] GetCharacter($x, $y)
    {
        if (($x -lt 0) -or ($x -ge ([Bricks]::windowWidth)))
        {
            return $null
        }

        if (($y -lt 0) -or ($y -ge ([Bricks]::windowHeight)))
        {
            return $null
        }

        $character = [Bricks]::cellToCharacter[$x, $y]
        if ($character -and $character.isFinished)
        {
            return $null
        }

        return $character
    }
}

class BricksCharacter : RenderItem
{
    BricksCharacter($cell, $cellTrailing, $x, $y)
        : base($cell, $cellTrailing, $x, $y)
    {
    }

    [void] Update($dt)
    {
    }
}

class BricksBall : RenderItem
{
    $velocity = @(0.0, 0.0)

    BricksBall($x, $y)
        : base((New-Object System.Management.Automation.Host.BufferCell(
            [char]0x2b24,
            [System.ConsoleColor]::Yellow,
            [Bricks]::backgroundColor,
            [System.Management.Automation.Host.BufferCellType]::Complete)),
            $null,
            $x, $y)
    {
        $this.SetVelocity(-1.0, -1.0)
    }

    [void] Update($dt)
    {
        # assume cell width and height ratio is 1:2
        $newX = $this.x + $this.velocity[0] * $dt
        $newY = $this.y + $this.velocity[1] * 0.5 * $dt

        $this.SetNewPos($newX, $newY)
    }

    [void] SetVelocity($x, $y)
    {
        $length = [Math]::Sqrt($x * $x + $y * $y);
        $this.velocity[0] = $x * [Bricks]::kBallSpeed / $length
        $this.velocity[1] = $y * [Bricks]::kBallSpeed / $length
    }

    [void] SetNewPos($newX, $newY)
    {
        $walls = @()
        $walls += $this.GetWalls($this.x, $newX, 0)
        $walls += $this.GetWalls($this.y, $newY, 1)
        $walls = $walls | Sort-Object -Property t

        $dx = $newX - $this.x
        $dy = $newY - $this.y
        $windowSize = [Bricks]::windowWidth, [Bricks]::windowHeight

        $iSearchCoord = [Int]$this.x, [Int]$this.y
        foreach ($wall in $walls)
        {
            $componentIndex = $wall.componentIndex
            $iSearchCoord[$componentIndex] = $wall.cellIndex

            $isHit = $false
            $barVelocityAndInfluence = $null
            if (($iSearchCoord[$componentIndex] -lt 0) -or ($iSearchCoord[$componentIndex] -ge $windowSize[$componentIndex]))
            {
                $isHit = $true
            }
            else
            {
                $character = [Bricks]::GetCharacter($iSearchCoord[0], $iSearchCoord[1])
                if ($character)
                {
                    $character.Finish()
                    [Bricks]::CreateExplosion($character, $true)
                    $isHit = $true
                }
                elseif ([Bricks]::bar.IsInside($iSearchCoord[0], $iSearchCoord[1]))
                {
                    $barVelocityAndInfluence = [Bricks]::bar.GetBallVelocityAndInfluence()
                    $isHit = $true
                }
            }

            if ($isHit)
            {
                $newVelocity = $this.velocity.Clone()
                $newVelocity[$componentIndex] = -$newVelocity[$componentIndex]
                if ($barVelocityAndInfluence)
                {
                    $barVelocityX = $barVelocityAndInfluence[0]
                    $barVelocityY = $barVelocityAndInfluence[1]
                    $barInfluence = $barVelocityAndInfluence[2]

                    $barVelocityY *= [Math]::Sign($newVelocity[1])

                    $newVelocity[0] = Lerp $newVelocity[0] $barVelocityX $barInfluence
                    $newVelocity[1] = Lerp $newVelocity[1] $barVelocityY $barInfluence
                }
                $this.SetVelocity($newVelocity[0], $newVelocity[1])

                $this.x = Limit ($this.x + $dx * $wall.t) 0 ([Bricks]::windowWidth-1)
                $this.y = Limit ($this.y + $dy * $wall.t) 0 ([Bricks]::windowHeight-1)
                return
            }
        }

        $this.x = $newX
        $this.y = $newY
    }

    [Object[]] GetWalls($currentCoord, $newCoord, $componentIndex)
    {
        $walls = @()
        $min = [Math]::Min($currentCoord, $newCoord)
        $max = [Math]::Max($currentCoord, $newCoord)
        $d = $newCoord - $currentCoord
        $sign = [Math]::Sign($d)

        $wallCoords = @(([Math]::Floor($min))..([Math]::Floor($max)))
        if ($wallCoords.Count -gt 1)
        {
            $wallCoords = $wallCoords[1..($wallCoords.Count-1)]
            foreach ($wallCoord in $wallCoords)
            {
                $t = ($wallCoord - $currentCoord) / $d
                $walls += @{componentIndex = $componentIndex; t = $t; cellIndex = ($wallCoord + $sign)}
            }
        }
        return $walls
    }
}

class BricksBar
{
    $x = 0
    $y = 0
    $width = 0
    $renderItems = $null
    $nextMoveDirection = 0
    $prevMoveDirection = 0
    $velocityInfluence = 0.0

    BricksBar($x, $y, $width)
    {
        $this.x = $x
        $this.y = $y
        $this.width = $width

        $color = [System.ConsoleColor]::Blue
        $firstCell = New-Object System.Management.Automation.Host.BufferCell(
            "(",
            $color,
            [Bricks]::backgroundColor,
            [System.Management.Automation.Host.BufferCellType]::Complete)

        $lastCell = New-Object System.Management.Automation.Host.BufferCell(
            ")",
            $color,
            [Bricks]::backgroundColor,
            [System.Management.Automation.Host.BufferCellType]::Complete)

        $middleCell = New-Object System.Management.Automation.Host.BufferCell(
            " ",
            [Bricks]::foregroundColor,
            $color,
            [System.Management.Automation.Host.BufferCellType]::Complete)

        $this.renderItems = New-Object System.Collections.ArrayList
        $this.renderItems.Add([RenderItem]::new($firstCell, $null, $x, $y)) | Out-Null
        for ($i = 1; $i -lt ($width-1); ++$i)
        {
            $this.renderItems.Add([RenderItem]::new($middleCell, $null, ($x + $i), $y)) | Out-Null
        }
        $this.renderItems.Add([RenderItem]::new($lastCell, $null, $x + ($width-1), $y)) | Out-Null
    }

    [void] Render($windowBuffer)
    {
        foreach ($item in $this.renderItems)
        {
            $item.Render($windowBuffer)
        }
    }

    [void] Update($dt)
    {
        if ($this.nextMoveDirection -eq 0)
        {
            $this.velocityInfluence = 0.0
        }
        elseif (($this.prevMoveDirection * $this.nextMoveDirection) -lt 0)
        {
            # Moved in opposite direction
            $this.velocityInfluence = 0.0
        }
        else
        {
            $this.velocityInfluence += $dt / [Bricks]::kBarVelocityInfluenceAccelTime
            $this.velocityInfluence = [Math]::Min($this.velocityInfluence, 1.0)
        }

        $this.x += $this.nextMoveDirection * [Bricks]::kBarSpeed * $dt
        $this.prevMoveDirection = $this.nextMoveDirection
        $this.nextMoveDirection = 0

        $maxX = [Bricks]::windowWidth - $this.width - 1
        if (($this.x -lt 0) -or ($this.x -gt $maxX))
        {
            $this.velocityInfluence = 0.0
        }
        $this.x = [Math]::Max($this.x, 0)
        $this.x = [Math]::Min($this.x, $maxX)

        for ($i = 0; $i -lt $this.width; ++$i)
        {
            $this.renderItems[$i].x = $this.x + $i
        }
    }

    [void] MoveLeft()
    {
        $this.nextMoveDirection = -1
    }

    [void] MoveRight()
    {
        $this.nextMoveDirection = 1
    }

    [Boolean] IsInside($x, $y)
    {
        if ($y -ne $this.y)
        {
            return $false
        }

        if ($x -lt $this.x)
        {
            return $false
        }

        if ($x -ge ($this.x + $this.width))
        {
            return $false
        }

        return $true
    }

    [Double[]] GetBallVelocityAndInfluence()
    {
        $velocityX = [Math]::Sign($this.prevMoveDirection) * [Bricks]::kBallSpeed
        $velocityY = 0.5 * [Bricks]::kBallSpeed
        return $velocityX, $velocityY, $this.velocityInfluence
    }
}


class Explosion : RenderItem
{
    $timer = 0.0
    $colorAnimation = $null
    $isBackgroundColorAnimation = $true
    $propagationDelayTime = 0.0

    static $kColors = @(
        [System.ConsoleColor]::Yellow,
        [System.ConsoleColor]::Yellow,
        [System.ConsoleColor]::Blue,
        [System.ConsoleColor]::DarkBlue
    )

    Explosion($character, $isBackgroundColorAnimation)
        : base(
            $character.cells[0],
            $character.cells[1],
            $character.x, $character.y)
    {
        $this.colorAnimation = [LinearAnimation]::new([Explosion]::kColors, [Bricks]::kExplosionDuration)
        $this.isBackgroundColorAnimation = $isBackgroundColorAnimation
        $this.propagationDelayTime = [Bricks]::kExplosionPropagationDelayTime.Get()
    }

    [void] Update($dt)
    {
        if ($this.isFinished)
        {
            return
        }

        $this.colorAnimation.Update($dt)
        $color = $this.colorAnimation.GetValue()
        if ($this.isBackgroundColorAnimation)
        {
            $this.SetCell($null, $null, $color)
        }
        else
        {
            $this.SetCell($null, $color, $null)
        }

        if ($this.colorAnimation.GetRatio() -gt $this.propagationDelayTime)
        {
            foreach ($offsetX in @(-1..1))
            {
                foreach ($offsetY in @(-1..1))
                {
                    $character = [Bricks]::GetCharacter($this.x + $offsetX, $this.y + $offsetY)
                    if ($character)
                    {
                        $character.Finish()
                        [Bricks]::CreateExplosion($character, $false)
                    }
                }
            }
        }

        if ($this.colorAnimation.IsFinished())
        {
            $this.Finish()
        }
    }
}

PSClearHost/Private/Falling.ps1
using module .\Common.psm1

class Falling
{
    static $characters = $null
    static $startLine = 0
    static $windowHeight = 0

    static $kStartLineVelocity = 0.2
    static $kStartTime = [RandomDouble]::new(0.0, 0.1)
    static $kMass = [RandomDouble]::new(1.0, 6.0)
    static $kColorAnimDuration = [RandomDouble]::new(0.8, 1.2)
    static $kColorAnimStartTime = 0.6
    static $kGravity = 9.8
    static $kDragCoef = 10.0
    static $kMeterToCellRatio = 20

    static [void] Init([System.Management.Automation.Host.BufferCell[,]]$windowBuffer)
    {
        [Falling]::windowHeight = $windowBuffer.GetUpperBound(0)
        [Falling]::characters = CreateCharacters ([FallingCharacter]) $windowBuffer

        $maxY = 0
        foreach ($character in [Falling]::characters)
        {
            $maxY = [Math]::Max($character.y, $maxY)
        }
        [Falling]::startLine = $maxY
    }

    static [void] Term()
    {
        [Falling]::characters = $null
    }

    static [Boolean] Render([System.Management.Automation.Host.BufferCell[,]]$windowBuffer, [Double]$dt, [Double]$speed)
    {
        [Falling]::startLine -= [Falling]::kMeterToCellRatio * [Falling]::kStartLineVelocity * $speed * $dt
        foreach ($character in [Falling]::characters)
        {
            if ($character.y -gt [Falling]::startLine)
            {
                $character.Start()
            }
            $character.Update($speed * $dt)
        }

        $isAllFinished = $true
        foreach ($character in [Falling]::characters)
        {
            if (-not $character.isFinished)
            {
                $isAllFinished = $false
                break
            }
        }
        if ($isAllFinished)
        {
            return $true
        }

        foreach ($character in [Falling]::characters)
        {
            $character.Render($windowBuffer)
        }
        return $false
    }
}

class FallingCharacter : RenderItem
{
    $velocity = 0.0
    $mass = 1.0
    $startTime = 0
    $timer = 0
    $colorAnimation = $null
    $colorAnimStartTime = 0
    $isStarted = $false

    FallingCharacter($cell, $cellTrailing, $x, $y)
        : base($cell, $cellTrailing, $x, $y)
    {
        $this.startTime = [Falling]::kStartTime.Get()
        $this.mass = [Falling]::kMass.Get()

        $colorAnimDuration = [Falling]::kColorAnimDuration.Get()

        $this.colorAnimation = [LinearAnimation]::new(@(
            [System.ConsoleColor]::Blue,
            [System.ConsoleColor]::Blue,
            [System.ConsoleColor]::Blue,
            [System.ConsoleColor]::DarkBlue,
            [System.ConsoleColor]::Green
        ), $colorAnimDuration)

        $this.colorAnimStartTime = $this.startTime + [Falling]::kColorAnimStartTime
    }

    [void] Start()
    {
        $this.isStarted = $true
    }

    [void] Update($dt)
    {
        if (-not $this.isStarted)
        {
            return
        }
        if ($this.isFinished)
        {
            return
        }

        $this.timer += $dt
        if ($this.timer -lt $this.startTime)
        {
            return
        }

        $a = [Falling]::kGravity - ([Falling]::kDragCoef * $this.velocity / $this.mass)
        $this.velocity += $a * $dt
        $this.y += [Falling]::kMeterToCellRatio * $this.velocity * $dt

        if ($this.y -ge [Falling]::windowHeight)
        {
            $this.Finish()
        }

        if ($this.timer -gt $this.colorAnimStartTime)
        {
            $this.colorAnimation.Update($dt)
            $color = $this.colorAnimation.GetValue()

            $this.SetCell($null, $color, $null)
        }
    }
}

PSClearHost/Public/Clear-HostRaindrop.ps1
# . "$PSScriptRoot\..\Private\Common.psm1"
# . "$PSScriptRoot\..\Private\Bricks.ps1"
# . "$PSScriptRoot\..\Private\Falling.ps1"
# . "$PSScriptRoot\..\Private\Flipping.ps1"

$players = @{
    Falling  = [Falling]
    Flipping = [Flipping]
    Bricks   = [Bricks]
}

Function Clear-HostRaindrop {
    <#
    .SYNOPSIS
        Clears the host and displays a raindrop animation.
    .DESCRIPTION
        Clears the host and displays a raindrop animation.
    .PARAMETER Mode
        Type of Animation to use; If not set, the animation type will be randomly selected.
        Possible Modes are `Falling`, `Flipping`, or `Bricks`.
    .PARAMETER Speed
        The playback speed of the raindrop animation.
    .PARAMETER Framerate
        The framerate of the raindrop animation.
    .PARAMETER DebugSkipRender
        Skips rendering the animation to the host.
    .INPUTS
        None
    .OUTPUTS
        None
    .EXAMPLE
        Clear-HostRaindrop -Mode 'Flipping' -Speed 0.5

        Clears the host and displays a Flipping raindrop animation with a playback speed of 0.5x.
    .EXAMPLE
        Clear-HostRaindrop

        Clears the host and displays a raindrop animation.
    .EXAMPLE
        Clear-HostRaindrop -Speed 1.5 -Framerate 60

        Clears the host and displays a raindrop animation with a speed of 1.5 and a framerate of 60.
    .EXAMPLE
        Clear-HostRaindrop -DebugSkipRender

        Clears the host and displays a raindrop animation, but skips rendering the animation to the host.
    .LINK
        https://github.com/jimbrig/PSClearHost
    #>

    [CmdletBinding()]
    param (
        [Parameter(ValueFromPipelineByPropertyName = $true)]
        [ValidateSet('Falling', 'Flipping', 'Bricks')]
        [String]$Mode = ($players.Keys | Get-Random),

        [Parameter(ValueFromPipelineByPropertyName = $true)]
        [ValidateRange(0.1, 10)]
        [Double]$Speed = 1.0,

        [Parameter()]
        [Switch]$DebugSkipRender
    )

    Begin {
        # $dt = 1.0 / $Framerate
        $vsyncTimer = New-Object System.Diagnostics.StopWatch
        $dtTimer = New-Object System.Diagnostics.StopWatch

        $vsyncTimer.Start()
        $dtTimer.Start()
    }

    Process {

        if ($null -eq $host.UI.RawUI.WindowSize) {
            # RawUI is not available (e.g. Windows PowerShell ISE).
            Clear-Host
            throw 'RawUI is not available (e.g. Windows PowerShell ISE).'
        }

        if ($DebugSkipRender) {
            return
        }

        Play $players[$Mode] $Speed


        # $windowSize = $host.UI.RawUI.WindowSize
        # $windowPosition = $host.UI.RawUI.WindowPosition
        # $windowWidth = $windowSize.Width
        # $windowHeight = $windowSize.Height



    }

    End {
        $vsyncTimer.Stop()
        $dtTimer.Stop()

        [Console]::CursorVisible = $prevCursorVisible
    }

}

New-Alias -Name clear -Value Clear-HostRaindrop -Option AllScope -Force
New-Alias -Name cls -Value Clear-HostRaindrop -Option AllScope -Force


#     $player = $playerClass::New($windowRect, $speed)
#     $player.Init()
#
#     $vsyncTimer.Start()
#     $dtTimer.Start()
#     while (-not $player.IsDone())
#     {
#         $dtTimer.Stop()
#         $dt = $dtTimer.Elapsed.TotalSeconds
#         $dtTimer.Reset()
